In [0]:
import tensorflow as tf
mnist = tf.keras.datasets.mnist

In [0]:
(x_train, y_train),(x_test, y_test) = mnist.load_data()
x_train = x_train / 255.0
x_test = x_test / 255.0

In [0]:
x_train.shape   # 공부하는 과정이니 이런거 확인해보기 

(60000, 28, 28)

In [0]:
x_test.shape

(10000, 28, 28)

In [0]:
#모델 구조 설정
model = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(input_shape=(28, 28)),
  #tf.keras.layers.Dense(512, activation=tf.nn.relu),  # 이걸 활성화시키면 더 정확해짐. 히든 노드를 둠으로써 정확도가 올라간다. 이 위에 3줄이 가장 중요 렐루 의미알기
  tf.keras.layers.Dense(256, activation=tf.nn.relu),
  tf.keras.layers.Dense(10, activation=tf.nn.softmax)  #소프트맥스는 시그모이드랑 비슷  이 구조 잘봐두기 dense라는 거미줄이 우리가 늘 보던 구조임.
])



In [0]:
model.summary()   #Dense라는 표현 알아두기 layer의 타입이 dense 나중에 우리 dense없앨거야

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten_1 (Flatten)          (None, 784)               0         
_________________________________________________________________
dense_2 (Dense)              (None, 256)               200960    
_________________________________________________________________
dense_3 (Dense)              (None, 10)                2570      
Total params: 203,530
Trainable params: 203,530
Non-trainable params: 0
_________________________________________________________________


In [0]:
784 * 10 + 10

7850

In [0]:
784*512 + 512

401920

In [0]:
512*10 + 10

5130

In [0]:
401920 + 5130

407050

params 가 왜 7850개 일까? 전체 파라미터도 7850개고 컴퓨터가 학습할 수 있는 파라미터도 7850개네  이게 뭔뜻이지?
784 * 10 + 10 인것 같은데 그게 무슨 의미지?

In [0]:
 # 모델 푸는 방법 지정
  model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])
  

In [0]:
#풀자! 
model.fit(x_train, y_train, epochs=5)  #epoch를 더 많이 돌리면 좀 더 정확해지겠지 그리고  히든레이어를 두면 더 정확해질거임/ 이건 사람이 정했으니 하이퍼파라미터 5의 의미는 한 epoch당 60000개의 데이터를 한번 다쓰는것. 5니까 그걸 5번돌린다. 여기서의 accuracy는 train데이터의 정확도임. 봤던데이터로 하니까 정확도


Epoch 1/5
60000/60000 [==============================] - 8s 138us/sample - loss: 0.2264 - acc: 0.9340
Epoch 2/5
60000/60000 [==============================] - 8s 134us/sample - loss: 0.0923 - acc: 0.9722
Epoch 3/5
60000/60000 [==============================] - 8s 134us/sample - loss: 0.0618 - acc: 0.9808
Epoch 4/5
60000/60000 [==============================] - 8s 136us/sample - loss: 0.0450 - acc: 0.9861
Epoch 5/5
60000/60000 [==============================] - 8s 135us/sample - loss: 0.0332 - acc: 0.9895


### *Real World Challenge*: Large difference between training and testing set accuracy

test accuracy와 training accuracy의 뜻도 알기 차이를 알아야함. 정확도가 얼마다라고 얘기할때 정확도를 보통 training accuracy로 얘기하는데 의미있는건 test accuracy

#### Test accuracy

In [0]:
model.evaluate(x_test, y_test)  # train 에 사용안했던 10000개 데이터 넣고 돌렸더니 정확도가 97%
# model.evaluate(x_test[:2], y_test[:2])  # 2개 데이터만 test해보고 싶을때.

10000/10000 [==============================] - 1s 53us/sample - loss: 0.0728 - acc: 0.9786


[0.07282146261757007, 0.9786]

#### Training accuracy

In [0]:
model.evaluate(x_train, y_train)  # train data로도 한번 테스트 해볼까?

60000/60000 [==============================] - 3s 45us/sample - loss: 0.0236 - acc: 0.9929


[0.023638827126581843, 0.99291664]

이 위 두개의 차이를 아는게 중요한데 test로 나온 정확도를 말해야지. 그런데 test데이터로 97나온걸 어떻게 더 올릴까 생각하기 전에 training데이터를 가지고 어떻게 더 정확도를 올릴지 먼저 고민해야지 봤던 것 자료부터 잘 맞추도록.  안봤던 데이터로 해도 잘나오지만 봤던 데이터로 정확도 나온게 당연히 더 높을 것